test/y2storage/filesystems/base_test.rb

```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/env rspec
+# encoding: utf-8
+
+# Copyright (c) [2017] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require_relative "../spec_helper"
+require "y2storage"
+
+describe Y2Storage::Filesystems::Base do
+
+  before do
+    fake_scenario(scenario)
+  end
+  let(:scenario) { "mixed_disks_btrfs" }
+  let(:blk_device) { Y2Storage::BlkDevice.find_by_name(fake_devicegraph, "/dev/sda2") }
+  subject(:filesystem) { blk_device.blk_filesystem }
+
+  describe "#free_space" do
+    context "#detect_space_info succeed" do
+      it "return #detect_space_info#free" do
+        fake_space = double(free: Y2Storage::DiskSize.MiB(5))
+        allow(filesystem).to receive(:detect_space_info).and_return(fake_space)
+
+        expect(filesystem.free_space).to eq Y2Storage::DiskSize.MiB(5)
+      end
+    end
+
+    context "#detect_space_info failed" do
+      before do
+        allow(filesystem).to receive(:detect_space_info).and_raise(Storage::Exception, "error")
+      end
+
+      context "it is block filesystem" do
+        context "detect_resize_info succeed" do
+          it "returns size minus minimum resize size" do
+            size = Y2Storage::DiskSize.MiB(10)
+            allow(filesystem).to receive(:detect_resize_info).and_return(double(min_size: size))
+
+            expect(filesystem.free_space).to eq(blk_device.size - size)
+          end
+        end
+
+        context "detect_resize_info failed" do
+          it "returns zero" do
+            size = Y2Storage::DiskSize.MiB(0)
+            allow(filesystem).to receive(:detect_resize_info).and_raise(Storage::Exception, "Error")
+
+            expect(filesystem.free_space).to eq size
+          end
+        end
+      end
+
+      context "it is not block filesystem" do
+        it "returns zero" do
+          allow(filesystem).to receive(:is?).and_return(false)
```

Comment:
Too much mocking. Wouldn't it be better to use a real Nfs object here?

Reply:
Which boils down to another example in which is obvious that the RSpec tests are written while reading the source code (same workflow). Without reading the structure of the `if` and `rescue` I would have written:

- For an NFS share
  - if it can be mounted (#detect_space_info succeeds)
  - if it cannot be mounted (#detect_space_info fails)

- For a BlkFilesystem
  - if it can be mounted (#detect_space_info succeeds)
  - if it cannot be mounted (#detect_space_info fails)
    - if #detect_resize_info succeeds
    - if #detect_resize_info fails

Because for me NFS and blk filesystems are completely different scenarios, no matter whether they share same code path.

Reply:
problem with your approach is that it makes hard to extend it. E.g. when there is third type of Filesystem, it will be completely ignore. Which I am not sure if it is good. So now after changing implementation maybe something like if filesystem have block devices or not, is determinant and not its type anymore.